### DISCUSSION TITLE
Dynamically replacing NFC data.

### DISCUSSION THREAD

**Discussion by xxx2451** [nfcgate/nfcgate#160] [Q&A] [answered]
I tried to implement it, but it doesn't work. is it possible?

**Comment by kleest** [comment]
Please be more specific, otherwise we cannot help you: What are you trying to achieve? What have you tried?

To dynamically replace data in the relay or replay mode, you have to implement a server plugin. Replacing data in the NFCGate Android app itself is not supported.

**Reply by xxx2451** [reply]
I tried to do package substitution in the application NFCGate for Android. learn how to replace packages using filters . If a card sends  AID  0x50, we want to change the AID to 0x51. where is it easier to implement, on the server or in the application?

**Reply by xxx2451** [reply]
you can show an example of how to implement this.

**Reply by kleest** [reply, marked as answer]
The Android app does not have support for changing NFC data. You can only implement this in the server by writing a "plugin". There is an example plugin [mod_log.py](https://github.com/nfcgate/server/blob/v2/plugins/mod_log.py). The server calls the `def handle_data(log, data, state)` of the enabled plugins. The method receives every data that is relayed over the server. As you can see in mod_log.py, the data is a protobuf message. To change NFC data it needs to be parsed. The modified data must then be encoded as a protobuf message again and returned from the function.
